odoo/tools/translate.py

```diff
@@ -887,107 +906,94 @@ def push_translation(module, type, name, id, source, comments=None):
         if not sanitized_term or len(sanitized_term) <= 1:
             return
 
-        tnx = (module, source, name, id, type, tuple(comments or ()))
-        to_translate.add(tnx)
+        self._to_translate.add((module, source, name, res_id, ttype, tuple(comments or ())))
+
+    def _get_translatable_record(self, model, res_id):
+        if model not in self.env:
+            _logger.error("Unable to find object %r", model)
+            return False
+
+        original = record = self.env[model].browse(res_id)
+        if not record.exists():
+            _logger.warning("Unable to find object %r with id %d", model, res_id)
+            return False
 
-    def translatable_model(record):
         if not record._translate:
             return False
 
         if record._name == 'ir.model.fields.selection':
             record = record.field_id
         if record._name == 'ir.model.fields':
             field_name = record.name
-            field_model = env.get(record.model)
+            field_model = self.env.get(record.model)
             if (field_model is None or not field_model._translate or
                     field_name not in field_model._fields):
                 return False
 
-        return True
-
-    query = 'SELECT min(name), model, res_id, module FROM ir_model_data'
+        return original
 
-    if 'all_installed' in modules:
-        query += ' WHERE module IN ( SELECT name FROM ir_module_module WHERE state = \'installed\') '
+    def export_translatable_records(self):
+        """ Export translations of all translated records having an external id """
 
-    if 'all' not in modules:
-        query += ' WHERE module IN %s'
-        query_param = (tuple(modules),)
-    else:
-        query += ' WHERE module != %s'
-        query_param = ('__export__',)
+        query = """SELECT min(name), model, res_id, module
+                     FROM ir_model_data
+                    WHERE {module_clause}
+                 GROUP BY model, res_id, module
+                 ORDER BY module, model, min(name)"""
 
-    query += ' GROUP BY model, res_id, module ORDER BY module, model, min(name)'
+        if 'all' not in self._modules:
+            query = query.format(module_clause="module IN %s")
+            query_param = (tuple(self._modules),)
+        else:
+            query = query.format(module_clause="module != %s")
+            query_param = ('__export__',)
 
-    cr.execute(query, query_param)
+        self._cr.execute(query, query_param)
 
-    for (xml_name, model, res_id, module) in cr.fetchall():
-        xml_name = "%s.%s" % (module, xml_name)
+        for (xml_name, model, res_id, module) in self._cr.fetchall():
```

Comment:
Note for later: we should group records by `model`, and fetch their translations in batch.